File: unity/Net/UDPChannel.cs
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

using System;
using System.IO;
using System.Linq;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Net;
using System.Net.Sockets;
using System.Net.Security;
using System.Net.NetworkInformation;
using System.Text;
using System.Runtime;
using System.Runtime.Serialization;

using Michitai.Lan;
using Michitai.Lan.Data;
using Michitai.Lan.Net;
using Michitai.Lan.Net.Multiplayer;
using Michitai.Lan.Net.Multiplayer.Chat;
using Michitai.Lan.Net.Multiplayer.Commands;
using Michitai.Lan.Net.Multiplayer.Data;
using Michitai.Lan.Debug;
using UnityEngine;

namespace Michitai.Lan.Net
{
    /// <summary>
    /// Unreliable UDP datagram channel for high-frequency state synchronization
    /// (positions, transforms, inputs) at 30-60+ Hz without TCP head-of-line blocking.
    /// Datagram format: [1-byte flags][payload], payload compressed when beneficial.
    /// </summary>
    public sealed class UDPChannel
    {
        /// <summary>
        /// Maximum UDP datagram payload (IPv4). Datagrams larger than this are rejected.
        /// </summary>
        public const int MaxDatagramSize = 65507;

        /// <summary>
        /// Datagrams larger than this may be fragmented by IP and lose parts on congested networks.
        /// Keep per-frame state under this size for reliability (Ethernet MTU 1500 minus headers).
        /// </summary>
        public const int SafeDatagramSize = 1472;

        private UdpClient _socket;

        private bool _closed;

        private readonly object _send_lock;

        private readonly List<IPEndPoint> _peers;

        private readonly object _peers_lock;

        /// <summary>
        /// Event raised when a datagram is received. First argument is the sender endpoint.
        /// </summary>
        public event Action<IPEndPoint, Message> OnReceive;

        /// <summary>
        /// Gets the local endpoint the channel is bound to, or null when not started.
        /// </summary>
        public IPEndPoint IPEndPoint
    {
        get
        {
            try
            {
                return _socket?.Client?.LocalEndPoint as IPEndPoint;
            }
            catch
            {
                return null;
            }
        }
    }

        /// <summary>
        /// Gets the endpoints this channel has received datagrams from. Thread-safe snapshot.
        /// Used by Broadcast to reach all known peers.
        /// </summary>
        public IPEndPoint[] Peers
    {
        get
        {
            lock (_peers_lock)
            {
                return _peers.ToArray();
            }
        }
    }

        /// <summary>
        /// Gets whether the channel is closed.
        /// </summary>
        public bool IsClosed => _closed;



        /// <summary>
        /// Initializes a new instance of UDPChannel bound to the specified endpoint.
        /// </summary>
        /// <param name="point">The local IP endpoint to bind to. Use port 0 for an ephemeral port.</param>
        public UDPChannel(IPEndPoint point)
    {
        _socket = new UdpClient(point);

        _peers = new List<IPEndPoint>();

        _peers_lock = new object();

        _send_lock = new object();

        _closed = false;
    }

        /// <summary>
        /// Initializes a new instance of UDPChannel bound to the specified IP and port.
        /// </summary>
        /// <param name="ip">The local IP address to bind to.</param>
        /// <param name="port">The local port to bind to. Use 0 for an ephemeral port.</param>
        public UDPChannel(IPAddress ip, int port) : this(new IPEndPoint(ip, port))
    {
    }

        /// <summary>
        /// Initializes a new instance of UDPChannel bound to a free port from the specified range.
        /// </summary>
        /// <param name="ip">The local IP address to bind to.</param>
        /// <param name="range">The port range to select a free port from.</param>
        public UDPChannel(IPAddress ip, PortRange range)
    {
        PortRange.Store store = range.RangeStore;

        while (true)
        {
            try
            {
                _socket = new UdpClient(new IPEndPoint(ip, store.RandomPort));

                break;
            }
            catch (SocketException)
            {
            }
        }

        _peers = new List<IPEndPoint>();

        _peers_lock = new object();

        _send_lock = new object();

        _closed = false;
    }



        /// <summary>
        /// Starts the asynchronous receive loop.
        /// </summary>
        public void Start()
    {
        try
        {
            _socket.BeginReceive(ReceiveCallback, null);
        }
        catch
        {
            Stop();
        }
    }

        /// <summary>
        /// Stops the channel and releases the socket.
        /// </summary>
        public void Stop()
    {
        if (_closed)
            return;

        _closed = true;

        OnReceive = null;

        try
        {
            _socket?.Close();

            _socket?.Dispose();
        }
        catch
        {
        }
    }



        /// <summary>
        /// Sends a message to the specified endpoint as a single datagram.
        /// </summary>
        /// <param name="target">The destination endpoint.</param>
        /// <param name="message">The message to send.</param>
        /// <exception cref="InvalidDataException">Thrown when the datagram exceeds MaxDatagramSize.</exception>
        public void Send(IPEndPoint target, Message message)
    {
        Send(target, message?.GetMessage ?? string.Empty);
    }

        /// <summary>
        /// Sends a string to the specified endpoint as a single datagram.
        /// </summary>
        /// <param name="target">The destination endpoint.</param>
        /// <param name="text">The string to send.</param>
        /// <exception cref="InvalidDataException">Thrown when the datagram exceeds MaxDatagramSize.</exception>
        public void Send(IPEndPoint target, string text)
    {
        byte[] payload = Encoding.UTF8.GetBytes(text ?? string.Empty);

        bool compressed = Compressor.TryCompress(payload, out payload);

        byte[] datagram = new byte[payload.Length + 1];

        datagram[0] = compressed ? Frame.FlagCompressed : Frame.FlagNone;

        Buffer.BlockCopy(payload, 0, datagram, 1, payload.Length);

        if (datagram.Length > MaxDatagramSize)
        {
            throw new InvalidDataException($"Datagram exceeds maximum size ({datagram.Length} > {MaxDatagramSize})");
        }

        if (datagram.Length > SafeDatagramSize)
        {
            DebugConsole.LogWarning($"[Michitai.Lan][UDP-CHANNEL][DATAGRAM-FRAGMENTATION-RISK][{datagram.Length}]");
        }

        lock (_send_lock)
        {
            if (_closed)
                return;

            _socket.Send(datagram, datagram.Length, target);
        }
    }

        /// <summary>
        /// Sends a message to all endpoints that have sent datagrams to this channel.
        /// </summary>
        /// <param name="message">The message to broadcast.</param>
        public void Broadcast(Message message)
    {
        IPEndPoint[] peers = Peers;

        for (int i = 0; i < peers.Length; i++)
        {
            try
            {
                Send(peers[i], message);
            }
            catch
            {
            }
        }
    }

        /// <summary>
        /// Removes an endpoint from the known peers list (e.g., when a client disconnects).
        /// </summary>
        /// <param name="point">The endpoint to forget.</param>
        public void Forget(IPEndPoint point)
    {
        lock (_peers_lock)
        {
            _peers.Remove(point);
        }
    }



        /// <summary>
        /// Callback for asynchronous datagram reception.
        /// </summary>
        /// <param name="result">The asynchronous result.</param>
        private void ReceiveCallback(IAsyncResult result)
    {
        IPEndPoint remote = new IPEndPoint(IPAddress.Any, 0);

        byte[] datagram;

        try
        {
            datagram = _socket.EndReceive(result, ref remote);
        }
        catch
        {
            if (!_closed)
            {
                Stop();
            }

            return;
        }

        if (datagram != null && datagram.Length > 1)
        {
            try
            {
                byte flags = datagram[0];

                int length = datagram.Length - 1;

                byte[] payload = new byte[length];

                Buffer.BlockCopy(datagram, 1, payload, 0, length);

                if ((flags & Frame.FlagCompressed) == Frame.FlagCompressed)
                {
                    payload = Compressor.Decompress(payload);
                }

                TrackPeer(remote);

                OnReceive?.Invoke(remote, new Message(Encoding.UTF8.GetString(payload)));
            }
            catch
            {
                DebugConsole.LogError("[Michitai.Lan][UDP-CHANNEL][RECEIVE-ERROR]");
            }
        }

        try
        {
            _socket.BeginReceive(ReceiveCallback, null);
        }
        catch
        {
            if (!_closed)
            {
                Stop();
            }
        }
    }

        /// <summary>
        /// Tracks a sender endpoint for later Broadcast use.
        /// </summary>
        /// <param name="point">The endpoint to track.</param>
        private void TrackPeer(IPEndPoint point)
    {
        lock (_peers_lock)
        {
            if (!_peers.Contains(point))
            {
                _peers.Add(point);
            }
        }
    }
}
}

File: cpp/src/Net/UDPChannel.cpp
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

#include <Michitai/Lan/Net/UDPChannel.hpp>
#include <Michitai/Lan/Net/Compression.hpp>
#include <Michitai/Lan/Net/Frame.hpp>
#include <Michitai/Lan/Debugging/DebugConsole.hpp>
#include "SocketInternal.hpp"

#include <algorithm>
#include <stdexcept>


namespace Michitai::Lan::Net
{
    UDPChannel::UDPChannel(const IPEndPoint& point)
    {
        _socket = std::make_unique<Detail::Socket>(Detail::Socket::UdpV4());

        if (!_socket->Bind(point))
            Detail::ThrowSocketError("bind");
    }

    UDPChannel::UDPChannel(const IPAddress& ip, const PortRange& range)
    {
        PortRange::Store store = range.RangeStore();

        while (true)
        {
            int port;
            try
            {
                port = store.RandomPort();
            }
            catch (const std::out_of_range&)
            {
                throw std::runtime_error("UDPChannel: no free port in range");
            }

            auto socket = std::make_unique<Detail::Socket>(Detail::Socket::UdpV4());
            if (socket->Bind(IPEndPoint(ip, port)))
            {
                _socket = std::move(socket);
                break;
            }
        }
    }

    UDPChannel::~UDPChannel()
    {
        Stop();
        JoinThreads();
    }

    std::optional<IPEndPoint> UDPChannel::IpEndPoint() const
    {
        try
        {
            return _socket ? _socket->LocalEndPoint() : std::nullopt;
        }
        catch (...)
        {
            return std::nullopt;
        }
    }

    std::vector<IPEndPoint> UDPChannel::Peers() const
    {
        std::lock_guard lock(_peersMutex);
        return _peers;
    }

    void UDPChannel::Start()
    {
        try
        {
            Ptr self = shared_from_this();
            _receiveThread = std::thread([self] { self->ReceiveLoop(); });
        }
        catch (...)
        {
            Stop();
        }
    }

    void UDPChannel::Stop()
    {
        if (_closed.exchange(true))
            return;

        OnReceive = nullptr;

        if (_socket)
            _socket->Close();
    }

    void UDPChannel::Send(const IPEndPoint& target, const Message& message)
    {
        Send(target, message.GetMessage());
    }

    void UDPChannel::Send(const IPEndPoint& target, const std::string& text)
    {
        std::vector<std::uint8_t> payload(text.begin(), text.end());

        const bool compressed = Compressor::TryCompress(payload, payload);

        std::vector<std::uint8_t> datagram(payload.size() + 1);
        datagram[0] = compressed ? Frame::FlagCompressed : Frame::FlagNone;
        std::copy(payload.begin(), payload.end(), datagram.begin() + 1);

        if (static_cast<int>(datagram.size()) > MaxDatagramSize)
        {
            throw std::length_error("Datagram exceeds maximum size (" +
                                    std::to_string(datagram.size()) + " > " +
                                    std::to_string(MaxDatagramSize) + ")");
        }

        if (static_cast<int>(datagram.size()) > SafeDatagramSize)
        {
            Debug::DebugConsole::LogWarning("[Michitai.Lan][UDP-CHANNEL][DATAGRAM-FRAGMENTATION-RISK][" +
                                            std::to_string(datagram.size()) + "]");
        }

        std::lock_guard lock(_sendMutex);

        if (_closed.load())
            return;

        _socket->SendTo(target, datagram.data(), static_cast<int>(datagram.size()));
    }

    void UDPChannel::Broadcast(const Message& message)
    {
        const std::vector<IPEndPoint> peers = Peers();

        for (const IPEndPoint& peer : peers)
        {
            try
            {
                Send(peer, message);
            }
            catch (...)
            {
            }
        }
    }

    void UDPChannel::Forget(const IPEndPoint& point)
    {
        std::lock_guard lock(_peersMutex);
        std::erase(_peers, point);
    }

    void UDPChannel::ReceiveLoop()
    {
        std::vector<std::uint8_t> buffer(MaxDatagramSize);

        while (!IsClosed())
        {
            IPEndPoint remote;
            const int received = _socket->ReceiveFrom(buffer.data(), static_cast<int>(buffer.size()), &remote);

            if (received <= 0)
            {
                if (_closed.load())
                    return;

                Stop();
                return;
            }

            if (received > 1)
            {
                try
                {
                    const std::uint8_t flags = buffer[0];

                    std::vector<std::uint8_t> payload(buffer.begin() + 1, buffer.begin() + received);

                    if ((flags & Frame::FlagCompressed) == Frame::FlagCompressed)
                    {
                        payload = Compressor::Decompress(payload);
                    }

                    TrackPeer(remote);

                    OnReceive.Invoke(remote, Message(std::string(payload.begin(), payload.end())));
                }
                catch (...)
                {
                    Debug::DebugConsole::LogError("[Michitai.Lan][UDP-CHANNEL][RECEIVE-ERROR]");
                }
            }
        }
    }

    void UDPChannel::TrackPeer(const IPEndPoint& point)
    {
        std::lock_guard lock(_peersMutex);

        if (std::find(_peers.begin(), _peers.end(), point) == _peers.end())
        {
            _peers.push_back(point);
        }
    }

    void UDPChannel::JoinThreads()
    {
        if (_receiveThread.joinable())
        {
            if (_receiveThread.get_id() == std::this_thread::get_id())
                _receiveThread.detach();
            else
                _receiveThread.join();
        }
    }
}
